torch/csrc/jit/codegen/cuda/lower_loops.cpp

```diff
@@ -88,137 +103,83 @@ void LoopNestGenerator::handle(const Expr* expr) {
-      // This view is the same as the last axis, so start where we left off.
-      start = last_ca_view_ind + 1;
-    } else {
-      // This is a new view, figure out where we are in it, and start from there
-      for (start = 0; start < ca_view->nDims(); start++) {
-        if (loop_structure.back() == ca_view->getComputeAtAxis(start).first) {
-          break;
-        }
-      }
-      start++;
-    }
+    // Safe to use loop map since this is outside the compute at point
+    auto concrete_id =
+        gpu_lower->caParallelMap().getConcreteMappedID(out_tv->axis(out_i));
+    loop_structure.push_back(concrete_id);
+  }
 
-    // Go from start, and open all loops in the computeAt view until we hit the
-    // one associated with out->getComputeAtAxis(out_i)
-    for (size_t ca_i = start; ca_i < ca_view->nDims(); ca_i++) {
-      // Note that ca_view->getComputeAtAxis(ca_i) is equivalent to
-      // std::pair(ca_view->axis(ca_i), ca_view)
-      loop_structure.push_back(ca_view->getComputeAtAxis(ca_i).first);
-
-      // Update the last view processed
-      last_ca_view_ind = ca_i;
-      last_ca_view = ca_view;
-      if (ca_view->getComputeAtAxis(ca_i).first == ca_id) {
-        break;
-      }
-    }
+  auto out_id_it = loop_structure.begin();
+  auto for_loop_it = for_loops_.begin();
+  auto last_for_loop_matched = for_loops_.begin();
+
+  // If the loop is not within the compute at point,
+  // Tee up the loop structure
 
-    // Shouldn't ever hit this, but make sure we hit the break above, meaning we
-    // added all necessary axes from the compute at view.
-    TORCH_INTERNAL_ASSERT(
-        ca_view->getComputeAtAxis(last_ca_view_ind).first == ca_id);
+  while (out_id_it != loop_structure.end() && for_loop_it != for_loops_.end()) {
+    auto lowered_out_id =
+        gpu_lower->lowerValue(*out_id_it)->as<kir::IterDomain>();
+    if (gpu_lower->caLoopMap().areMapped(
+            lowered_out_id, (*for_loop_it)->iter_domain())) {
+      out_id_it++;
+      last_for_loop_matched = ++for_loop_it;
+    } else {
+      ++for_loop_it;
+    }
   }
 
-  // We're up to the compute at point in loop_structure, grab the remaining
-  // axes.
-  for (int64_t out_i = (int64_t)out->getThisComputeAtAxis();
-       out_i < (int64_t)out->nDims();
+  // Append axes outside the computeAt to the loop structure
+  for (auto out_i = out_tv->getThisComputeAtAxis();
+       out_i < (unsigned int)out_tv->nDims();
        out_i++) {
-    // It's actually local, but getComputeAtAxis returns a std::pair, axis
-    // doesn't
-    loop_structure.push_back(out->getComputeAtAxis(out_i).first);
+    loop_structure.push_back(out_tv->axis((int)out_i));
   }
 
-  // At this point loop_structure contains our overal target loop nest structure
-  // Lets get a copy of the loop structure, and figure out which loops we need
-  // to open.
-  auto loops_to_open = loop_structure;
+  // Save position of out_id_it as we will append to loop structure
```

Comment:
Shouldn't this be placed before the above for loop (i.e., before line 148)?

Reply:
Yes, it should be.

Reply:
Moved.